### DISCUSSION TITLE
Help flashing Sonoff SPM-Main module

### DISCUSSION THREAD

**Discussion by Katoz96** [arendst/Tasmota#16565] [Q&A] [answered]
I have bought a Sonoff SPM-Main module and I'm trying to flash Tasmota on it but no matter what I do I always get the same errors.

First I tried with the Tasmota Web Installer but I kept getting the error `Failed to initialize. Try resetting your device or holding the BOOT button while clicking INSTALL.`

Then I tried ESP-Flasher and esptool.py but it didn't work either and I was get the error `Unexpected error: ESP Chip Auto-Detection failed: Failed to connect to Espressif device: No serial data received.` and `Unexpected error: could not open port 'COM11': PermissionError(13, 'Access is denied.', None, 5)` if I cliked the View Log button

I tried all this things but nothing has changed:

- I swapped the tx and rx wires countless times 

- I powered the board with main power (disconnecting the 3v wire)

- I reconnected the front leds to make sure it was entering in programming mode

- At first I was shorting the GPIO0 pin to GND only for a couple of seconds during boot but then I soldered the two pads together.

- I changed the serial adapter, I was using a AZDelivery FT232RL that I use only to flash Tasmota (I even leave the USB cable attached so I know that combination will work the next time that I need to flash Tasmota) and I tried to use and Arduino UNO as a serial adapter
- I Closed all the programs in case there was one that was using the serial port
- I tried running ESP-Flasher as an administrator 
- I used another computer
- I reinstalled the adapter's driver
- Changed the USB port
- I spent hours on Google looking for a solution
- I followed [this guide ](https://knowledge.ni.com/KnowledgeArticleDetails?id=kA03q000000YGw9CAG&l=en-US), the service value was FTSER2K searching for it on Process Explorer was not used by anyone.

I tried so many combinations and I don't know what to do anymore, I hope someone can help me

**Comment by sfromis** [comment]
Powering with mains while connecting flashing pins at the same time risks damaging the device, especially with energy monitoring hardware.

Not clear what specific wiring you have, common ground is essential.

Those FT232RL devices are usually not good for *reliable* flashing (especially with ESP32) as they can't deliver much 3.3V power.

AFAIK, Arduino UNO is a 5V board, another risk of hardware damage as ESP chips are at 3.3V level.

**Reply by Katoz96** [reply]
I saw a post from Arendst where he said that he used the AC main to power the module and use the gnd, tx, rx for the serial connection. 

As for Arduino, I used 3.3v pin to power the board.

**Reply by Jason2866** [reply]
If Theo is doing such stuff he uses optocouplers to isolate AC mains from USB (PC).

**Comment by Katoz96** [comment, marked as answer]
I bought a CH340G converter and it worked flawlessly the first time using the web installer

**Comment by lalo-uy** [comment]
If the USB converter do not give enough power you can use a separate 3,3v power supply.
